A Dogleg Trust Region Hybrid Nonlinear solver

rcarson3 opened the pull request:
This new solver is based on the algorithm from the HybrdJ solver available within MINPACK. I've currently tested the solver against the MINPACK solver, and a python implementation and they do seem to be returning consistent answers except for the one noted test case. Although, the one test case ended up being a valid solution as well. So, I'm currently chalking that difference up to slight differences in how things are numerically implemented, between the C++ version and my python version. I can supply the python version as well if there's interest in that as a test case.

As part of this work, I also added a linalg header/namespace that contains several common numerical matrix/vector operations we've used either here or in other codes such as ECMech. My hopes is that this new set of code can eventually be used in place of what's used in our other libraries, as I'd like to reduce the amount of repeated code we have for these common operations.

It also contains a Householder-based QR decomposition as well as Givens rotation helper function. Eventually, I plan on adding an actual QR solver in there as a potential alternative to the LU solver in the trust-region dogleg nonlinear solver if the user needs something a bit more robust. However, we might need to add a pivoted QR decomposition instead of the current version for those needs.

Finally, I've added a few test cases to test the new solver. One is the Broyden solver we've used in the `TrDLDenseG` solver, and the other is chebyquad equations listed in https://people.sc.fsu.edu/~jburkardt/f_src/test_nonlin/test_nonlin.html and #5. 

While this should work / compile on the GPU, I still need to actually test that before this PR can be merged.

- [ ] Test on GPU

rcarson3 committed:
Adding a version of the HybrdJ solver from MINPACK
This code is based on the algorithm from the HybrdJ solver avaible within MINPACK.

nrbarton commented on SNLS_NLSPowell.h:
```diff
@@ -0,0 +1,729 @@
+   >::type
+>: std::true_type { static constexpr bool value = true;};
+
+// Hybrid Jacobian method from MINPACK which is a modified Powell method
```
Are there copyright issues here? I am not familiar with the license for MINPACK, or with how similar the coding here might be to what is in MINPACK.

rcarson3 replied:
So, the algorithm itself would be fine, but we might need to see how things are originally defined in the original paper/conference proceedings in regards to the exit conditions of the algorithm. I've ordered the conference book as the chapter is impossible to find online..., so I'll know the answer to that sometime next week. I did also find the following MINPACK document that actually goes over all the various exit conditions and algorithms as well: http://cds.cern.ch/record/126569/files/CM-P00068642.pdf. I have done somethings differently, but the exit conditions are largely the same as what's defined by minpack...

Outside that it appears that MINPACK (http://www.netlib.org/minpack/disclaimer) has essentially a BSD type license, so it might be fairly simple to include this in the header but as per your later comment we should probably go through another release process for this just to make sure things are good.

rcarson3 replied:
With the last bit of refactoring, I would say we are quite a bit different than that we should no longer have to worry about any copyright issues. I've also removed all these sorts of references outside the one at the top that mentions that we used MINPACK, the original Powell paper, and our own implementation as inspiration for this solver.

nrbarton commented on SNLS_NLSPowell.h:
```diff
@@ -0,0 +1,729 @@
+    // which is based on the old MINPACK version of things
+    __snls_hdev__
+    inline
+    void dogleg(double* const x,
```
Eventually this is code that could be shared between hybrd and the solver in SNLS_TrDLDenseG.h? Perhaps not, but it might be worth adding a comment about how/why the two are different.

rcarson3 replied:
Since this makes uses portion of the R factorization from QR factorization, I don't think they would be reusable between the two solvers...

rcarson3 replied:
I've been going through this a bit more, and I actually think a lot of this code could be shared. I've gone through both codes, and they're honestly doing the exact same thing with some minor differences in how various portions are done. They would have some small differences here and there, but it should be doable over all.

rcarson3 replied:
They all use the same dogleg solver now.

nrbarton commented on SNLS_NLSPowell.h:
```diff
@@ -0,0 +1,729 @@
+        
+        // at m_x
+        // We currently don't check on the success of this...
+        // and we should at this back in at some point
```
These comments have typos and are potentially confusing. 

If computeRJ fails then you should do at least something, rather than continuing merrily on your way.

nrbarton commented on SNLS_linalg.h:
```diff
@@ -0,0 +1,339 @@
+/// where vec has length ndim
+template<int ndim>
+__snls_hdev__
+inline double norm(const double* const vec)
```
If you are going to add these kinds of things here, you might want to consider switching at least normvec and normvecSq in SNLS_TrDLDenseG.h to use the functions in SNLS_linalg.h

rcarson3 replied:
I can do that.

rcarson3 replied:
This change should now be done as part of all the refactoring.

nrbarton commented on SNLS_NLSPowell.h:
```diff
@@ -0,0 +1,729 @@
+//	have trait nDimSys
+//
+template< class CRJ, bool scaleDiag = false >
+class SNLSHybrdDenseG 
```
You have Powell, Hybrid, and Hybrd all used in the naming of this solver. To make it easier for others to find related content, I recommend you pick one and stick with it. Including in file names (this file and SNLS_testdriver_powell.cc).

rcarson3 replied:
I'll work on the naming issue to unify it across all the different cases.

rcarson3 replied:
The names should all be more or less consistent at this point. Although, we might still want to modify them to something else.

nrbarton commented on SNLS_qr_solve.h:
```diff
@@ -0,0 +1,204 @@
+// Algorithm taken from Matrix Computation 4th ed. by GH Golub and CF Van Loan
+// pg. 236 alg. 5.1.1
+__snls_hdev__
+inline
```
Is there an advantage to templating this on length, or do you expect inlining to make it as performant as it is likely to get?

rcarson3 replied:
The length is dynamic within the QR factorization, so we can't template this unfortunately. We do know it has a max value of ndim and a minimum value of 1 depending on what loop we're on in the QR factorization.

nrbarton commented on SNLS_linalg.h:
```diff
@@ -0,0 +1,339 @@
+printMat(const double* const A) {
+    for (int iX = 0; iX<n; ++iX) {
+        for (int jX = 0; jX<n; ++jX) {
+            std::cout << std::setw(21) << std::setprecision(14) << A[SNLS_NN_INDX(iX, jX, n)] << " ";
```
I do not see where SNLS_NN_INDX is coming from. It does not look like it is in SNLS_base.h

It is the same as SNLSTRDLDG_J_INDX? If so, why not combine them so that the same macro is used for both?

rcarson3 replied:
It turns out I forgot to upload the changes that include this value in `SNLS_base.h`. I'll add that back in as part of the PR refactor as well as some modifications to the Trust-Region Dogleg solver to make use of some of the linalg functions.

nrbarton commented on SNLS_testdriver_powell.cc:
```diff
@@ -0,0 +1,350 @@
+#include <cstdlib>
```
This seems excessively duplicative. If SNLS_NN_INDX and SNLSTRDLDG_J_INDX become the same, then it seems like you could use some compile-time settings to run the Broyden or ChebyQuad test problem either for the SNLSTrDlDenseG solver or for the SNLSHybrdDenseG solver. That is, if the idea is that this is a solver library and users can swap among solvers relatively readily, then you could set a good example of how that would look.

How about setting up a test problem per source file and then having compile-time settings for which solver gets used?

rcarson3 replied:
I think that should be easy enough to do. It will just require a little bit of refactoring of the test problems. I'll include this as part of some other refactoring.

nrbarton commented on SNLS_NLSPowell.h:
```diff
@@ -0,0 +1,729 @@
+
+/** Helper templates to ensure compliant CRJ implementations */
+template<typename CRJ, typename = void>
+struct nls_has_valid_computeRJ : std::false_type { static constexpr bool value = false;};
```
This is duplicative with has_valid_computeRJ in SNLS_TrDLDenseG.h. Unifying would enforce the idea that one can switch among solvers for a given problem. 

Same for nls_has_valid_computeRJ and other helper templates below.

nrbarton commented on SNLS_NLSPowell.h:
```diff
@@ -0,0 +1,729 @@
+    SNLSStatus_t solveInit(double* const residual)
+    {
+        m_jevals = 0;
+        m_fevals = 1; 
```
If you are always using this->computeRJ to do function and residual evaluations, then it might make more sense to initialize m_fevals to 0 and put the increment of m_fevals and m_jevals inside of this->computeRJ (with a conditional on the J being null).

rcarson3 replied:
I've thought of that, and I can't remember why I was avoiding it originally.

rcarson3 replied:
Made this change, but I still reset them to zero when this called.

nrbarton commented on SNLS_NLSPowell.h:
```diff
@@ -0,0 +1,729 @@
+            this->computeRJ(residual, nullptr);
+            // Need to think about what to do if this fails as things
+            // are handled differently from the TrDL solver.
+            // if ( !(rjSuccess) ) {
```
Repeated typos in comments, and the need to do _something_ if computeRJ comes back with an error. My guess is that you just go to SNLS_FAIL if computeRJ has an error, at least for now. You can code a cutback in the step size or the like later.

rcarson3 replied:
Yeah I wasn't too sure how best to handle the cut back portion. I can just add a SNLS_FAIL in the mean time until we figure that issue out.

rcarson3 replied:
This has been fixed with latest revisions of things.

nrbarton commented on SNLS_NLSPowell.h:
```diff
@@ -0,0 +1,729 @@
+                m_status = deltaFailure;
+                return m_status;
+            }
+            if (m_nslow2 == 5) {
```
Back to the copyright question -- this looks like the kind of detail that, if the same as in MINPACK or elsewhere, might trigger a need to be careful about release and licensing details.

rcarson3 replied:
So, these values are directly defined in the MINPACK manual. So, it would not be a copyright issue as we're just following their API. 

Additionally, I'm going to refactor the PR a bit as I've finally got my hands on the original conference proceeding's paper/chapter that this was algorithm was based on. I actually like the original a bit more as it shares some commonality with our trust-region dog-leg solver. My plans are therefore to do a bit of a hybrid where we just re-use our https://github.com/LLNL/SNLS/blob/develop/SNLS_TrDLDenseG.h#L37-L216 class here to control how delta grows and shrinks. Next, I'll keep the minpack logic for how to exit early if the solver is making no progress as well as their improvement on the original algorithm of doing a rank-1 update of the Jacobian using the Broyden method rather than what appears to be an update of a finite difference method given in the original paper.

nrbarton commented on SNLS_NLSPowell.h:
```diff
@@ -0,0 +1,729 @@
+
+        snls::linalg::matTVecMult<m_nDim, m_nDim>(Q, residual, qtf);
+
+        // std::cout << "R" << std::endl;
```
Either remove this dead code or put it in a debug ifdef?

rcarson3 replied:
Missed that when pushing the commits up. I'll remove that statement as it was used while debugging.

nrbarton commented on SNLS_NLSPowell.h:
```diff
@@ -0,0 +1,729 @@
+            }
+
+            // Compute the ratio of the actual to the predicted reduction
+            const double ratio = (predicted_reduction > 0.0) ? (actual_reduction / predicted_reduction) : 0.0;
```
If m_os != nullptr, the value of this ratio might be something to print. Or at least in the TrDLDenseG case it can be helpful -- but there one is always using the Jacobian from an evaluation, so maybe things are not as straightforward to interpret here. In any case, your output to m_os is rather sparse if someone does want to use that output to help diagnose problems.

rcarson3 replied:
I can add that information as an output if m_os is not null and we're in a debug build. I imagine another useful spot would be around the dogleg update potentially with the delta X increment and then after the rank1 update of the Q, R, and Q^T f values.

rcarson3 replied:
The refactor has eliminated this portion of code and now if the user has provided m_os then this sort of thing will be printed. In fact, I'm pretty sure we print all the same information as the `TrDLDenseG` solver now thanks to the fact we now share so much code.

rcarson3 committed:
Merge branch 'cmake_fixes' into hybrid_nonlinear_solver

rcarson3 committed:
Various build fixes and a few small bug fixes to get inline with v0.3.0

rcarson3 committed:
Merge branch 'develop' into hybrid_nonlinear_solver

rcarson3 committed:
Remove duplicate code in classes where possible and make use of linalg header

rcarson3 commented:
Currently I'm working on pulling out the dogleg calculation and sharing that between solvers. The nice thing about doing this is that I've realized there's some nice re-organization we could do for this part of the calculations which should reduce the number of calculations being done :)

edit: In https://github.com/LLNL/SNLS/pull/6/commits/f9225f010fe39cb27300853aaa25587e48342698 I refactored the dogleg / update delta x portion of the trust region solver. My plans are to use a similar version of things in a universal dog leg version of the code. The commit should also show how many of the various outside loop variables we were able to eliminate which should be beneficial for the batch solvers.

edit2: In https://github.com/LLNL/SNLS/pull/6/commits/9bcd69a9fe8883e178607d7b5be341e25cd40ece I moved the dogleg stuff all over to its own function / header. Using the ExaCMech miniapp, I'm seeing a roughly 1-2% performance improvement over the traditional method which isn't a bad win. Next steps are to work on the hybrid solver and the batch trust region solver.

edit3: In https://github.com/LLNL/SNLS/pull/6/commits/21b5f9e8dce6eee75e248ca47577fdf26a455adf I finally got all the batch dogleg stuff working. In order to do this, I did need to shuffle some stuff around to make all of the testing framework and etc happy when the batch framework is activated. The nice thing about this reshuffling of things is our batch memory usage drops by a decent amount as we now require 7 fewer temp variables to carry around :)

rcarson3 committed:
Refactored dogleg / update delx formulation for NDim solvers

rcarson3 committed:
Move dogleg algo over to own func/header

rcarson3 committed:
Refactor batch dogleg portion of code
This refactor has also allowed to remove quite a few temporary variables that we no longer need

rcarson3 committed:
Reshuffling of typedefs and functions for batch/raja perf suite stuff
The compiler was not at all happy when we had raja perf suite active with our regular test suite. Hopefully, the rearrangment here fixes the issue.

rcarson3 committed:
Few more small bug fixes for batch dogleg...

rcarson3 commented:
Once I've finished with refactoring everything, it might be worth seeing if we want to create a base class ND solver class (base batch version as well) which the various classes can build upon. I've started to notice that the existing ones at least share a common set of features which could be carried over. I also imagine that this would help simplify our testing suite and would allow people to plug and play different solvers into their codes.

edit: I would say that if we did do the base class thing that this would be a breaking change which would require another version bump to 0.4.0 as the api changes.

nrbarton commented:
I just poked briefly through the recent round of changes. That is rather significant code restructuring. It would be worth checking that the behavior remains as desired, particularly for challenging problems that produce reductions in the delta and otherwise exercise a variety of code paths. If you let me know once you have done a bunch of testing and things have calmed down with the changes then I can try to take another look at the review.

Looking around now, there are still some references to MINPACK that might indicate copyright issues. If the coding in here is as in MINPACK in a significant way, then that could be an issue. If the coding is based on more general considerations that are shared by the MINPACK implementation then you probably want to adjust the comments to more clearly indicate that. 

I have not taken a look at the base class considerations, but agree the breaking changes probably warrant a version bump as you indicate.

nrbarton commented on src/SNLS_TrDLDenseG.h:
```diff
@@ -197,239 +196,99 @@ class SNLSTrDlDenseG
+            // This is done outside this step so that these operations can be done with varying solve
+            // techniques such as LU/QR or etc...
+            if(!reject_prev) {
+               // So the LU solve does things in step which causes issues...
```
I am not sure what "does things in step" means here. 

You are talking about the LU solve doing things in place, so that a standard matrix-vector multiple with the Jacobian no longer works?

rcarson3 replied:
Yeah sorry should have worded this better. Yes, the LU solve is done in-place which causes issues when doing normal matvec type operations.

rcarson3 commented:
> I just poked briefly through the recent round of changes. That is rather significant code restructuring. It would be worth checking that the behavior remains as desired, particularly for challenging problems that produce reductions in the delta and otherwise exercise a variety of code paths. If you let me know once you have done a bunch of testing and things have calmed down with the changes then I can try to take another look at the review.

So, the code works for all of our test and all of ECMech's as well as well as the miniapp. I do have a few challenging problems not in those test cases that I can further check things out with. However, our test cases do hit all the different paths based on their print outputs, and I'm quite confident that we won't see any changes. I'm still restructuring things, but I'll tag you once things are ready for another look-over. The last bit that I need to factor out of the classes is the update delta portion of things, but I don't expect there to be any algorithmic changes there.

> 
> Looking around now, there are still some references to MINPACK that might indicate copyright issues. If the coding in here is as in MINPACK in a significant way, then that could be an issue. If the coding is based on more general considerations that are shared by the MINPACK implementation then you probably want to adjust the comments to more clearly indicate that.
> 

So, I haven't gotten to that portion of the code yet to refactor. I've been working on creating a shared set of kernels between the solvers first. The end result is that the new solver is largely different from the MINPACK version. It'll honestly have inspirations from the original paper, the MINPACK version, and finally our existing solver. The few small pieces we might share in common are the early exit conditions if the Jacobian/solver is not making significant progress over a number of steps. These exit conditions are described in detail in their published API document, which I will make sure to cite in the code as well as the other papers as well.

rcarson3 committed:
Rename dogleg headers and move update delta functions over there
Delta update portion can be shared between solvers so move it over to a common header

rcarson3 committed:
Few small bug fixes related to batch version of things

rcarson3 committed:
First stab at completely refactoring hybrid solver (not compiled yet)
Have not compiled yet so no doubt there'll still be compiler errors however this should be approaching the final version. I need to update all the documentation and some of the testing and testing first though.

rcarson3 committed:
Code compiles now but test fails for hybrid solver

rcarson3 committed:
Small bug fixes, lots of debug info, and hybrid solver now kinda works
At least 3 of the tests at least now work but still have way more iterations than the previous version of things. ChebyQuad test problems still fail quite badly which might be due to the trust region changing the delta to aggressively

rcarson3 committed:
Clean-up of code, add documentations, change params so tests pass
Updated documentation to cite our various sources that we used to implement portions of the hybrid solver
Removed all the  old debug statements in the hybrid solver
Updated some params of the test suite so things start passing / solving

rcarson3 committed:
Rename hybrid solver files

rcarson3 committed:
Fix things due to name changes

rcarson3 commented:
@nrbarton the refactoring is more or less done at this point. I would say you should be able to start re-reviewing it at this point. The MINPACK solver and this one now behave quite differently, and we even see different convergence rate between the two. I think there's only a small fraction of similar things between the MINPACK version and our own at this point. First, I like their idea of exiting early if the solver doesn't appear to be making any progress, and so I'm making use of their API details to implement something similar in our own. Second, we both use a broyden update method to update the Jacobian which shouldn't be a concern as that method has very favorable features and is well defined in literature.

Outside that, I would say this is a pretty hefty PR with quite a lot of major refactoring to share a ton of code between solvers. It also adds in a linear algebra namespace, QR factorizations,  a new solver, and some new tests. We could probably do a bit more refactoring as well and combine the tests into one file as well and just have an ifdef on the test suites that are run in one go.

While reviewing I would say let me know if you'd like to change the name of anything as well. I changed the solvers name to something similar to our already existing ND solver, but we could always name it to something else.

One last thing, I did notice that our trust region can be brutal on how the delta is controlled between iterations of the solver which can cause some nonlinear systems to just not solve. I noticed that one of my previous tests for the new solver was no longer solving due to this change (It fails on our normal solver already, so it's not an issue with the new solver). One thing I'm tempted to do is have this new solver templated on a class that controls the delta so in the future it would be super simple to swap in a new delta control type class.

edit: I forgot to mention I still need to test this with some of my application codes that gave the trouble in the 1st place, but I expect things to still work more or less over there as well.

nrbarton commented:
Regarding 
> One last thing, I did notice that our trust region can be brutal on how the delta is controlled between iterations of the solver which can cause some nonlinear systems to just not solve. I noticed that one of my previous tests for the new solver was no longer solving due to this change (It fails on our normal solver already, so it's not an issue with the new solver). One thing I'm tempted to do is have this new solver templated on a class that controls the delta so in the future it would be super simple to swap in a new delta control type class.

... I am not sure what you mean by the delta control being brutal. The default settings might not be well-suited to a given problem, and things might work better if the `_xi*` control parameters in `TrDeltaControl` are adjusted. Although for severely nonlinear systems, some settings working better than others could be a matter of luck to some degree, rather than the settings being universally better for that class of problem. 

I am not averse to generalizing the step size control capability, as long as we are careful about code usability and performance. Although I also worry that it might give people something else to tinker with when a real issue is that they need to think about reformulating their problem setup to have more attractive numerical properties, or about debugging their Jacobian calculation.

nrbarton commented on src/SNLS_port.h:
```diff
@@ -19,8 +19,11 @@
 #include <exception>
 #include <stdexcept>
 #define SNLS_FAIL(loc,str) throw std::runtime_error(std::string("at ") + std::string(loc) + std::string(" failure : ") + std::string(str)) ;
+#define SNLS_WARN(loc, str) printf("WARNING : SNLS warning in %s : %s\n", loc, str);
```
It seems like you also need to define `SNLS_WARN` up above for cases in which `HAVE_MSLIB` is true

rcarson3 replied:
I'll add that one as well.

nrbarton commented on src/SNLS_linalg.h:
```diff
@@ -0,0 +1,364 @@
+/// The values are stored in norm_vec which is of length m
+template<int ndim, int mdim>
+__snls_hdev__
+inline void normColumn(const double* const matrix, 
```
Some of these functions (`normColumn`, `outerprod`, `outerProdScale`) do not seem to be used. While it might be convenient to shove them in here, I am not sure that it is the best approach. Is the idea that this will grow to become a more general linear algebra package distinct from the needs of SNLS?

rcarson3 replied:
Yes, the general thought was that we would be to have a general linear algebra set of functions that we could use in other libraries such as ECMech and elsewhere, so we can reduce duplicate code used in our various libraries.

nrbarton commented on src/SNLS_linalg.h:
```diff
@@ -0,0 +1,364 @@
+/// mat has dimensions ndim x mdim
+template <int ndim, int mdim>
+__snls_hdev__
+inline void outerprod(const double* const vec1,
```
Should be `outerProd` to be consistent with camelCase elsewhere.

nrbarton commented on src/SNLS_linalg.h:
```diff
@@ -0,0 +1,364 @@
+
+/// Takes the norm of the columns of a matrix
+/// where the matrix has dimensions nxm
+/// This assumes that matrix is ordered in row-major order
```
I imagine all of the matrices are assumed to be in row-major storage. It is odd to have a comment to that effect only for this function. Maybe move the comment to the top? Or add it consistently for all functions that take matrices?

rcarson3 replied:
I'll add a comment at the top of this header.

nrbarton commented on src/SNLS_linalg.h:
```diff
@@ -0,0 +1,364 @@
+                        const double* const a,
+                        double* const p)
+{
+    for (int iM = 0; iM < mdim; ++iM) {
```
Would it be better to flip the order of the loops (being careful about the initialization of p to zero) so that the inner loop has unit stride?

rcarson3 replied:
So, I noticed that after I made this change I saw a drop in performance (~7% drop) in the ExaCMech miniapp when running a batch version of the miniapp compiled with nvcc 10.1.243. The normal miniapp compiled with 11.4.1 didn't seem to care one way or the other how this formed. So, it's either less of an issue with newer nvcc compilers or just not a factor with the non-batch version...

The CPU version also seems to not really be affected by things. So, I'm going to revert the changes I made here, so the order of the loops isn't flipped.

nrbarton commented on src/SNLS_linalg.h:
```diff
@@ -0,0 +1,364 @@
+{
+    if (zero_out)
+    {
+        for (int iLM = 0; iLM < (ldim * mdim); iLM++) { prod[iLM] = 0.0; }
```
Here (for `ldim*mdim`) and elsewhere, the idea is that the compiler is smart enough to optimize and pull the multiply out of the loop?

rcarson3 replied:
Given these are templated values the compiler should be able to optimize these values out along with any dead code.

nrbarton commented on src/SNLS_TrDLDenseG.h:
```diff
@@ -30,42 +32,8 @@ extern "C" {
 
 //////////////////////////////////////////////////////////////////////
```
On line 47 (gitlab will not let me add a comment to lines that you did not change ...) "nad" should be "and"

nrbarton commented on src/SNLS_TrDLDenseG.h:
```diff
@@ -197,239 +165,62 @@ class SNLSTrDlDenseG
+               // if we want to keep this similar to the batch version of things
+               snls::updateDelta<_nDim>(_deltaControl, residual, res_0, pred_resid, nr_norm, _tolerance, use_nr, rjSuccess,
+                                        _delta, _res, _rhoLast, reject_prev, _status, _os);
+               // This new check is required due to moving all the delta update stuff into its own function to share features between codes
```
The comment referencing a "new" check here is only useful with reference to the old version of the code. So perhaps it is helpful for the review, but moving forward it could be confusing. I suggest modifying the comment to document how the code behaves without reference to previous behavior, or removing the comment.

Also, do we want a convention for how many columns are allowed? This comment goes out to column 139 ...

rcarson3 replied:
We currently don't have any formatting conventions within SNLS. We could just copy the uncrustify file used within in ExaCMech temporarily. I was actually thinking this might be something we might want to add.

nrbarton commented on src/SNLS_TrDLDenseG.h:
```diff
@@ -616,7 +357,7 @@ class SNLSTrDlDenseG
          for ( int iX=0; iX<_nDim; ++iX) {
             for ( int jX=0; jX<_nDim; ++jX) {
-               oss << std::setw(21) << std::setprecision(11) << A[SNLSTRDLDG_J_INDX(iX,jX,_nDim)] << " " ;
+               oss << std::setw(21) << std::setprecision(11) << A[SNLS_NN_INDX(iX,jX,_nDim)] << " " ;
```
Given `printVec` and `printMat` in SNLS_linalg.h, it seems like `printVecX` and `printMatJ` should go away and those functions from SNLS_linalg.h should be used instead.

nrbarton commented on src/SNLS_kernels.h:
```diff
@@ -0,0 +1,242 @@
+         }
+#ifdef SNLS_DEBUG
+         if ( beta > 1.0 || beta < 0.0 ) {
+            SNLS_FAIL(__func__, "beta not in [0,1]") ;
```
Do you really want this to error out for a debug build if you are expecting (line 146 below) to bring it back into [0,1] to deal with roundoff?

rcarson3 replied:
This was the default behavior from the original `TrDLDenseG` solver. If we don't want this behavior I can modify this.

nrbarton replied:
I would be tempted to at least change the ifdef so that it does not get activated by default for debug builds.

nrbarton commented on src/SNLS_kernels.h:
```diff
@@ -0,0 +1,242 @@
+      } // if norm_s_sd_opt >= delta
+   } // use_nr
+
+   // update x here as we want to avoid additional kernel calls
```
Does this comment about kernel calls really make sense here in the non-batch version?

rcarson3 commented:
> 
> ... I am not sure what you mean by the delta control being brutal. The default settings might not be well-suited to a given problem, and things might work better if the `_xi*` control parameters in `TrDeltaControl` are adjusted. Although for severely nonlinear systems, some settings working better than others could be a matter of luck to some degree, rather than the settings being universally better for that class of problem.
> 
> I am not averse to generalizing the step size control capability, as long as we are careful about code usability and performance. Although I also worry that it might give people something else to tinker with when a real issue is that they need to think about reformulating their problem setup to have more attractive numerical properties, or about debugging their Jacobian calculation.

I guess when I say brutal, I mean that when the solver starts having a hard time the delta can decrease rapidly enough that the system has a hard time moving back towards the actual solution. I noticed this with the `chebyquad function where n=3` which gives the solvers all kinds of trouble.  I've played around with the `_xiDecDelta` a bit to  see if I could find something to work, but it might just be that I needed to mess around with the other parameters as well. The reason I brought up allowing different methods to control the step size is that I noticed that the rather simplistic approach that MINPACK took to controlling it actually worked surprisingly well across a range of problems... So, it seems to me that it probably wouldn't hurt to give more advance users the ability some control over how this step size is calculated. However, we could maybe add this as a feature in a later point release. In the mean time, we could look at adding some documentation to the Trust region delta class in regards to how each parameter affects the solver?

nrbarton commented on src/SNLS_kernels_batch.h:
```diff
@@ -0,0 +1,255 @@
+   {
+      // Update the delta kernel
+      // this breaks out of the internal lambda and is essentially a loop continue
+      if( status[i + offset] != SNLSStatus_t::unConverged){ return; }
```
This return keeps the solver from continuing to update the solution once it has converged? That is a good feature to help prevent decomposition-based (happenstance of points ending up together in a batch) dependence in the results. It might be good to more clearly document that such is what is happening here.

rcarson3 replied:
Yes that is what this return does. I thought I had this explained at some point with the batch solver, but I guess that comment was lost in the wind. I'll add it back in.

nrbarton commented on src/SNLS_kernels_batch.h:
```diff
@@ -0,0 +1,255 @@
+template<int nDim>
+__snls_host__
+inline
+void updateDelta(const int offset,
```
It looks like offset is only used in a subset of the arrays. It might be worth documenting which use offset.

nrbarton commented on src/SNLS_TrDLDenseG_Batch.h:
```diff
@@ -125,13 +88,12 @@ class SNLSTrDlDenseG_Batch
       // 3d arrays [_initial_batch_size, CRJ::nDimSys, CRJ::nDimSys]
       // _Jacobian
-      const int num_allocs = _initial_batch_size * (11 + (5 * CRJ::nDimSys) + CRJ::nDimSys * CRJ::nDimSys) + 
+      const int num_allocs = _initial_batch_size * (4 + (4 * CRJ::nDimSys) + CRJ::nDimSys * CRJ::nDimSys) +
```
Would it help with code maintainability/documentation to use macros instead of the various (2, 3, 4) integer values that are sprinkled around in this coding? 

As an example, and something that should in any case be fixed, the comment at line 262 below looks like it is consistent with the old num_allocs calculation (with 11 and 5) rather than with the updated num_allocs calculation (with 4 and 4).

rcarson3 replied:
Macros aren't a bad idea. I'll modify this when I get a chance.

nrbarton commented on src/SNLS_TrDLDenseG_Batch.h:
```diff
@@ -304,23 +266,9 @@ class SNLSTrDlDenseG_Batch
          // if batch_size and npts are equal
```
At around lines 130, 212, and 244 : _status and _fevals are initialized three separate times. All three initializations are really necessary? And other things (_delta, _res, _x) do not need to be initialized in all three places?

rcarson3 replied:
So, we should keep the one in 130 as part of the class initialization just so users don't try to read from uninitialized memory. However, it seems like we can at least get rid of 212.

We should probably keep 244 as well as a user might decide to run a solver multiple times with either different options or a different state. So, we want to make sure those values are reset. Also, _delta and _res are initialized with proper values further later down so that's fine as well. We expect the user to have already set _x to something if they don't want it to have the default value of 0.0.

However, you are right that not all of them appear necessary. We can definitely get rid of the one at 212.

nrbarton commented on src/SNLS_TrDLDenseG_Batch.h:
```diff
@@ -331,8 +279,6 @@ class SNLSTrDlDenseG_Batch
-         // offset += off2d;
 
          int m_fevals = 0;
          int m_nJFact = 0;
```
At line 292, the comment should read "the last block" instead of "the last bulk"

nrbarton commented on src/SNLS_TrDLDenseG_Batch.h:
```diff
@@ -790,18 +527,6 @@ class SNLSTrDlDenseG_Batch
-      }
       
    public:
 
```
Code would be easier to read if the #ifdef etc at lines 543, 558 559, 574 were left-justified

nrbarton commented on src/SNLS_TrDLDenseG_Batch.h:
```diff
@@ -932,7 +657,7 @@ class SNLSTrDlDenseG_Batch
          oss << std::setprecision(14) ;
```
Comment reading
>       /// Performs a bitwise reduction and operation on _status to see if the
>      /// current batch can exit.

at lines 594 and 595 looks like cut-and-paste from the comment for `status_exit` without appropriate editing for `reduced_rjSuccess`

nrbarton commented on src/SNLS_TrDLDenseG_Batch.h:
```diff
@@ -932,7 +657,7 @@ class SNLSTrDlDenseG_Batch
          for ( int iX=0; iX<_nDim; ++iX) {
             for ( int jX=0; jX<_nDim; ++jX) {
-               oss << std::setw(21) << std::setprecision(11) << A[SNLSTRDLDG_J_INDX(iX,jX,_nDim)] << " " ;
+               oss << std::setw(21) << std::setprecision(11) << A[SNLS_NN_INDX(iX,jX,_nDim)] << " " ;
```
The `printVecX` and `pringMatJ` functions can be replaced by `printVec` and `printMat` from SNLS_linalg.h, such that the functions here can be eliminated?

nrbarton commented on src/SNLS_HybrdTrDLDenseG.h:
```diff
@@ -0,0 +1,599 @@
+// 		TODO ... J becomes a RAJA::View ?
+//	have trait nDimSys
+//
+// Might want to look into having this templated on a class that controls the delta as well as the trust region can be a bit aggressive in how it determines such things
```
Clarification of what you mean by aggressive might be useful here. 

And the line length is rather long.

nrbarton commented on src/SNLS_HybrdTrDLDenseG.h:
```diff
@@ -0,0 +1,599 @@
+
+        this->setOutputlevel( outputLevel );
+
+        m_complete = true;
```
Would it be worth adding this kind of machinery with m_complete in the other solver classes?

rcarson3 replied:
We kinda have a check like this in the other code by checking if `_deltaControl` (private member var) is set, but I can modify the other libraries to copy this feature.

nrbarton replied:
Whatever you think is best.

nrbarton commented on src/SNLS_HybrdTrDLDenseG.h:
```diff
@@ -0,0 +1,599 @@
+    inline bool computeRJ(double* const r,
+                          double* const J ) 
+    {
+        m_fevals += ((J == nullptr) ? 1 : 0);
```
The other solvers increment fevals for any computeRJ call, right? It would be good to have this solver behave in a manner that is consistent with the others.

rcarson3 replied:
So, this does act the same in principal. The only time we call `computeRJ` in this solver with `J != nullptr` is when we need to update the Jacobian. We actually don't care about the residual outputted at that point in time because we already have it from either `solveInit` or from within the `for(;;)` portion of `solveStep`.

nrbarton replied:
Understood. I guess it comes down to what we want fevals to track -- is it the number of times the function evaluation happens or the number of times that the solver needs the function evaluation. The computeRJ is set up to always compute the residual and to optionally compute the Jacobian. And I was interpreting fevals as the number of evaluations rather than the number of times an evaluation is actually needed. So in that meaning of fevals I think it would get incremented with every computeRJ call.

rcarson3 replied:
Okay if that's how you're interpreting things than I can have it increment every time.

nrbarton commented on src/SNLS_HybrdTrDLDenseG.h:
```diff
@@ -0,0 +1,599 @@
+        // we're essentially starting over here so we can reset these values
+        bool reject_prev = false;
+        double Jg_2 = 0.0;
+        double res_0 = m_res;
```
Given that `computeRJ` does not compute `m_res` it looks like you are counting on `m_res` having been computed from the last `updateDelta` call (or from having been reset to `res_0`). It might be worth adding a comment to clarify that.

rcarson3 replied:
I can add a comment just to make it's clear where it's set. Since, it's updated in `updateDelta` and initially set in `solveInit`.

nrbarton commented on src/SNLS_HybrdTrDLDenseG.h:
```diff
@@ -0,0 +1,599 @@
+            double pred_resid;
+            bool use_nr = false;
+
+            // If step was rejected nrStep will be the same value and so we can just recalculate it here
```
I do not understand this comment -- you are recalculating `nr_norm` whether or not the previous step was rejected -- but that is not really what your comment says. There is an equivalent comment in SNLS_TrDLDenseG.h that I also do not understand.

rcarson3 replied:
I see what you're talking about now. I'll clarify what I mean by `it` in this comment.

nrbarton commented on src/SNLS_HybrdTrDLDenseG.h:
```diff
@@ -0,0 +1,599 @@
+            // if we want to keep this similar to the batch version of things
+            snls::updateDelta<m_nDim>(m_deltaControl, residual, res_0, pred_resid, nr_norm, m_tolerance, use_nr, rjSuccess,
+                                     m_delta, m_res, m_rhoLast, reject_prev, m_status, m_os);
+            // This new check is required due to moving all the delta update stuff into its own function to share features between codes
```
See comment in SNLS_TrDLDenseG.h about changing this comment to not reference old code.

nrbarton commented on src/SNLS_HybrdTrDLDenseG.h:
```diff
@@ -0,0 +1,599 @@
+            reject_prev = false;
+            //
+            bool rjSuccess = this->computeRJ(residual, nullptr) ; // at _x
+            // Could also potentially include the reject previous portion in here as well
```
It might be better to put this comment inside of the `updateDelta` function, instead of repeating it both here and in SNLS_TrDLDenseG.h

nrbarton commented on src/SNLS_HybrdTrDLDenseG.h:
```diff
@@ -0,0 +1,599 @@
+            double temp = rmat[SNLS_NN_INDX(i, i, m_nDim)];
+            // If R has a tiny diagonal diagonal 
+            if (fabs(temp) < tol) {
+                // We could try something like the above to make sure this function always
```
You mean "like the coding below" instead of "like the above"?

rcarson3 replied:
Oops yep I meant "like the coding below"...

nrbarton commented on src/SNLS_HybrdTrDLDenseG.h:
```diff
@@ -0,0 +1,599 @@
+    // This performs a Broyden Rank-1 style update for Q, R and Q^T f
+    // This version has origins in this paper:
+    // Gill, Philip E., et al. "Methods for modifying matrix factorizations." Mathematics of computation 28.126 (1974): 505-535.
+    // However, you can generally find it described in more approachable manners else where on the internet
```
"else where" should be "elsewhere"

nrbarton commented on src/SNLS_qr_solve.h:
```diff
@@ -0,0 +1,204 @@
+
+namespace snls {
+
+// Need to add an actual QR solver in here at some point. It currently just contains just QR factorization functions
```
Change "just contains just" to "contains just"

nrbarton commented on src/SNLS_qr_solve.h:
```diff
@@ -0,0 +1,204 @@
+// pg. 236 alg. 5.1.1
+__snls_hdev__
+inline
+void houseHolderVec(double& beta,
```
Given that Householder is all one word, here and elsewhere you probably do not want to capitalize the "Holder" part of householder when doing camelCase.

nrbarton commented on test/SNLS_testdriver_hybrd.cc:
```diff
@@ -0,0 +1,368 @@
+};
+
+template<int nd>
+class ChebyQuad
```
Is there a reference or other documentation to be added for this test problem?

rcarson3 replied:
I do have a few references for this. I actually found the original paper this was defined in, but I found this in a nonlinear solver test suite elsewhere that I can also reference.

nrbarton commented on test/SNLS_testdriver_hybrd.cc:
```diff
@@ -0,0 +1,368 @@
+
+};
+
+// This test will fail due to our delta being too aggressive which causes the solution to not move 
```
Changes to _xiLG and _xiLO in the TrDeltaControl might allow the test to pass, depending on the nature of the "too aggressive" behavior that you reference.

rcarson3 replied:
Didn't really have any luck with at least the parameters I tried :/ I'm just going to leave this in as a known failed test, and if we end up allowing classes to be templated on a delta control type class then we can work on getting this one to pass.

nrbarton commented on test/SNLS_testdriver_hybrd.cc:
```diff
@@ -0,0 +1,368 @@
+{
+   const int nDim = ChebyQuad<5>::nDimSys;
+
+   ChebyQuad<5> chebyquad; // LAMBDA_BROYDEN 
```
Here and elsewhere, what is the point of the comment with `LAMBDA_BROYDEN` by the `chebyquad` declaration?

rcarson3 replied:
This was a copy and paste of test suites. I'll update this and remove that comment.

nrbarton reviewed (approved):
I finished another round of looking through the changes. Mostly it looks good. I will go ahead and mark it approved assuming that you will take appropriate action on the various comments.

rcarson3 committed:
Addressing various PR comments

rcarson3 commented:
@nrbarton the various comments should now be addressed. Some of the things I'll address in future PRs are working on the test structure so like tests are together, a formatting PR where we make use of either clang or uncrustify to enforce a style throughout, and finally I'm probably going to look more into variations of the update delta type stuff.

Also before I merge this, I'm going to do one more round of testing things out in ExaCMech using more production style runs to make sure everything still works as expected with all the refactoring that's gone on.

rcarson3 committed:
Small bug fix related to copy and paste error

rcarson3 committed:
Fix a missing ifdef for cuda builds and revert a loop ordering change
The loop order change had a negative impact on performance for batch solves. In a realistic usecase, I noticed a 7% drop in performance from this change, but only when running on the GPU. The CPU so no performance change.
Next, I was missing an ifdef which was causing build issues with GPU codes.

rcarson3 committed:
Missed this in last commit but copy and paste error for macro define value

rcarson3 committed:
hdev attributes were in wrong locations for qr funcs

rcarson3 commented:
So, I did one last round of testing with some GPU builds that also utilized the batch versions of things and tested in ExaCMech and its miniapp. I also did a round of a debug build/testing in ExaCMech and tested it on some tougher nonlinear systems which worked before all the refactoring. I caught a few small things that needed fixing. However, it appears to all be good now. 

I've also run the new solver in ExaCMech as a test, and it's working in there and returning  answers that appear the same as far as what's the miniapp printing. Initial testing resulted in a 2x slowdown, but I had changed the `_xiDecDelta` to 0.6 as this had worked with the Broyden tests for the new solver. I'll probably need to mess around with that as this could probably be moved closer to its original value...

@nrbarton thanks for reviewing it and suggesting various changes. I know this was a pretty beefy PR with some major refactoring and a new solver added. However, I think we're now in a pretty nice place for any new solvers or if we want to make our existing ones more flexible.

nrbarton commented:
> Initial testing resulted in a 2x slowdown, but I had changed the `_xiDecDelta` to 0.6 as this had worked with the Broyden tests for the new solver.

Thanks @rcarson3 for the follow-up. I would be interested to hear more about the 2x slowdown if you are able to pin that down. Definitely something to sort out. I would not expect a change to `_xiDecDelta` to make that much difference.
